### PR TITLE
Tf 2.0 od api models with loop

### PR DESCRIPTION
**Pull request by lazarevevgeny**
**Root cause analysis:** There are two issues:

- TensorFlow 2.4 Object Detection API models have new structure of the pre-processing block which is responsible for image resize and apply of mean/scale values. In a new version the pre-processing block is moved into the While (Loop) operation. MO does not handle new structure of the pre-processing block and fails to convert the model.
- There are two MO transformations `KerasRNNInputSlicing` and `KerasRNNOutputConcatenation` related to TF 2.X While operations which do not work properly for the TF 2.X While operations with 2+ inputs which are sliced or 2+ outputs being concatenated over each iteration. In this model the While operation with pre-processing block has 2 inputs and 2 outputs so `KerasRNNInputSlicing` and `KerasRNNOutputConcatenation` transformations crash. 

**Solution:** The solution for the issue number 1. There are multiple transformations enabled with the json configuration file to convert TF OD API models. The transformation `ObjectDetectionAPIPreprocessor2Replacement` for the pre-processor block is modified to find a pre-processing nodes (sequence of mul/add/div/sub operations) with constant inputs. Then:

- If the pre-processing nodes are located in the Loop then re-construct the pre-processing sub-graph in the main loop and remove operation performing resize.
- If the pre-processing nodes are located in the main graph then just remove operation performing resize.

The solution for the issue number 2.

- Created 2 anchor transformations `ObjectDetectionAPITransformationsStart` and `ObjectDetectionAPITransformationsFinish` and added dependencies between all TF OD API transformations to run them in between 
- Run graph clean up after `ObjectDetectionAPITransformationsFinish` to remove all dangling vertices which may be not connected after applying graph transformations (including the Loop operation with the pre-processing).
- Added dependency to `KerasRNNInputSlicing` and `KerasRNNOutputConcatenation` to run them after the TF OD API transformations.

Also, created transformation configuration files for 2.4 version of the models with updated names of the nodes surrounding image resize operation.

Ticket: 50525, 50635

Code:

* [x] Comments
* [x] Code style (PEP8)
* [x] Transformation generates reshape-able IR: N/A
* [x] Transformation preserves original framework node names: N/A because the mean/scale nodes being kept are moved from the internal graph

Validation: 

* [x] Unit tests: Done
* [x] Framework operation tests: N/A because no new operation enabled
* [x] Transformation tests: Done
* [x] Model Optimizer IR Reader check: Done

Documentation:

* [x] Supported frameworks operations list: N/A because no new operations enabled
* [x] Guide on how to convert the **public** model: minor update of the documentation with information on how to convert the models. Added a table with supported pre-trained TF 2.X OD API models.  The bigger updated will done as part of ticket 51541
* [x] User guide update: the major documentation update is needed but it will be done as part of a dedicated ticket

New structure of transformations dependencies:
transformations_dependencies

### CODE REVIEW

**Comment by rkazants** [comment]
What TensorFlow 2 layer has multiple input slicing and output concatenation pattern in While loop? Maybe you had a chance to look at training code.

File: docs/MO_DG/prepare_model/convert_model/Convert_Model_From_TensorFlow.md [hunk 1]
```diff
@@ -68,6 +68,43 @@ Detailed information on how to convert models from the <a href="https://github.c
 |Faster R-CNN Inception ResNet V2 Low Proposals Open Images\*|  [faster_rcnn_inception_resnet_v2_atrous_lowproposals_oid_2018_01_28.tar.gz](http://download.tensorflow.org/models/object_detection/faster_rcnn_inception_resnet_v2_atrous_lowproposals_oid_2018_01_28.tar.gz)|
 |Faster R-CNN ResNet 101 AVA v2.1\*|  [faster_rcnn_resnet101_ava_v2.1_2018_04_30.tar.gz](http://download.tensorflow.org/models/object_detection/faster_rcnn_resnet101_ava_v2.1_2018_04_30.tar.gz)|
 
+**Supported Pre-Trained Topologies from TensorFlow 2 Object Detection Models Zoo**
+
+Detailed information on how to convert models from the <a href="https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf2_detection_zoo.md">TensorFlow 2 Detection Model Zoo</a> is available in the [Converting TensorFlow Object Detection API Models](tf_specific/Convert_Object_Detection_API_Models.md) chapter. The table below contains models from the Object Detection Models zoo that are supported.
```
Comment (rkazants) [inline docs/MO_DG/prepare_model/convert_model/Convert_Model_From_TensorFlow.md:73]: Does it refer to the same chapter as for TensorFlow 1 OD models? Are any relevant specifics of TensorFlow 2 OD models there that must be documented?

File: docs/MO_DG/prepare_model/convert_model/tf_specific/Convert_Object_Detection_API_Models.md [hunk 2]
```diff
@@ -18,27 +18,30 @@ To convert a TensorFlow\* Object Detection API model, go to the `<INSTALL_DIR>/d
...
-    * `ssd_support_api_v.1.14.json` --- for frozen SSD topologies trained using the TensorFlow\* Object Detection API version 1.14 up to 1.14.X inclusively
-    * `ssd_support_api_v.1.15.json` --- for frozen SSD topologies trained using the TensorFlow\* Object Detection API version 1.15 up to 2.0
-    * `ssd_support_api_v.2.0.json` --- for frozen SSD topologies trained using the TensorFlow\* Object Detection API version 2.0 or higher
-    * `faster_rcnn_support.json` --- for frozen Faster R-CNN topologies from the models zoo
+    * `ssd_support_api_v.1.14.json` --- for SSD topologies trained using the TensorFlow\* Object Detection API version 1.14 up to 1.14.X inclusively
+    * `ssd_support_api_v.1.15.json` --- for SSD topologies trained using the TensorFlow\* Object Detection API version 1.15 up to 2.0
+    * `ssd_support_api_v.2.0.json` --- for SSD topologies trained using the TensorFlow\* Object Detection API version 2.0 up to 2.3.X inclusively
+    * `ssd_support_api_v.2.4.json` --- for SSD topologies trained using the TensorFlow\* Object Detection API version 2.4 or higher
```
Comment (rkazants) [inline docs/MO_DG/prepare_model/convert_model/tf_specific/Convert_Object_Detection_API_Models.md:24]: Did you check nightly build for TF 2.5: https://pypi.org/project/tf-nightly-cpu/

File: docs/MO_DG/prepare_model/convert_model/tf_specific/Convert_Object_Detection_API_Models.md [hunk 3]
```diff
@@ -18,27 +18,30 @@ To convert a TensorFlow\* Object Detection API model, go to the `<INSTALL_DIR>/d
...
+    * `efficient_det_support_api_v.2.0.json` --- for EfficientDet topologies trained using the TensorFlow\* Object Detection API version 2.0 up to 2.3.X inclusively
+    * `efficient_det_support_api_v.2.4.json` --- for EfficientDet topologies trained using the TensorFlow\* Object Detection API version 2.4 or higher
+    * `faster_rcnn_support.json` --- for Faster R-CNN topologies from the TF 1.X models zoo trained with TensorFlow\* version 1.9.0 or lower.
```
Comment (rkazants) [inline docs/MO_DG/prepare_model/convert_model/tf_specific/Convert_Object_Detection_API_Models.md:27]: There may be some unclarity between  `faster_rcnn_support.json` and `faster_rcnn_support_api_v1.7.json`.

File: docs/MO_DG/prepare_model/convert_model/tf_specific/Convert_Object_Detection_API_Models.md [hunk 4]
```diff
@@ -18,27 +18,30 @@ To convert a TensorFlow\* Object Detection API model, go to the `<INSTALL_DIR>/d
...
     * `faster_rcnn_support_api_v1.7.json` --- for Faster R-CNN topologies trained using the TensorFlow\* Object Detection API version 1.7.0 up to 1.9.X inclusively
     * `faster_rcnn_support_api_v1.10.json` --- for Faster R-CNN topologies trained using the TensorFlow\* Object Detection API version 1.10.0 up to 1.12.X inclusively
     * `faster_rcnn_support_api_v1.13.json` --- for Faster R-CNN topologies trained using the TensorFlow\* Object Detection API version 1.13.X
     * `faster_rcnn_support_api_v1.14.json` --- for Faster R-CNN topologies trained using the TensorFlow\* Object Detection API version 1.14.0 up to 1.14.X inclusively
     * `faster_rcnn_support_api_v1.15.json` --- for Faster R-CNN topologies trained using the TensorFlow\* Object Detection API version 1.15.0 up to 2.0
     * `faster_rcnn_support_api_v2.0.json` --- for Faster R-CNN topologies trained using the TensorFlow\* Object Detection API version 2.0 or higher
-    * `mask_rcnn_support.json` --- for frozen Mask R-CNN topologies from the models zoo
+    * `mask_rcnn_support.json` --- for Mask R-CNN topologies from the TF 1.X models zoo trained with TensorFlow\* version 1.9.0 or lower.
```
Comment (rkazants) [inline docs/MO_DG/prepare_model/convert_model/tf_specific/Convert_Object_Detection_API_Models.md:34]: `mask_rcnn_support.json` has interception with `mask_rcnn_support_api_v1.7.json` by TF version. Not clear which file to select then.

File: model-optimizer/mo/utils/class_registration.py [hunk 5]
```diff
@@ -128,7 +128,7 @@ def dump_graph_for_graphviz(self, node_attrs: list = [], edge_attrs: list = [],
         if nodes_to_dump is None:
             nodes_to_dump = self.nodes()
         string = '\ndigraph {\n'
-        string += 'node [color=lightblue2, style=filled];\n'
+        string += 'node [color=lightblue2, style=filled shape="box"];\n'
```
Comment (rkazants) [inline model-optimizer/mo/utils/class_registration.py:131]: Is comma missed?

**Comment by rkazants** [comment]
Also, I haven't found dependency to KerasRNNInputSlicing and KerasRNNOutputConcatenation to run them after the TF OD API transformations.

File: docs/MO_DG/prepare_model/convert_model/tf_specific/Convert_Object_Detection_API_Models.md (hunk 2 above)
Comment (lazarevevgeny) [reply docs/MO_DG/prepare_model/convert_model/tf_specific/Convert_Object_Detection_API_Models.md:24]: This is a common statement. In some cases the same config works for future versions, so that is why we don't set the upper bound.
I have not checked with 2.5.

File: docs/MO_DG/prepare_model/convert_model/tf_specific/Convert_Object_Detection_API_Models.md (hunk 3 above)
Comment (lazarevevgeny) [reply docs/MO_DG/prepare_model/convert_model/tf_specific/Convert_Object_Detection_API_Models.md]: Yes, thanks.

File: docs/MO_DG/prepare_model/convert_model/tf_specific/Convert_Object_Detection_API_Models.md (hunk 4 above)
Comment (lazarevevgeny) [reply docs/MO_DG/prepare_model/convert_model/tf_specific/Convert_Object_Detection_API_Models.md:34]: Fixed. Thanks.

File: docs/MO_DG/prepare_model/convert_model/Convert_Model_From_TensorFlow.md (hunk 1 above)
Comment (lazarevevgeny) [reply docs/MO_DG/prepare_model/convert_model/Convert_Model_From_TensorFlow.md:73]: Nothing special. Only different configuration files.

File: model-optimizer/mo/utils/class_registration.py (hunk 5 above)
Comment (lazarevevgeny) [reply model-optimizer/mo/utils/class_registration.py]: The code works and without comma. But I add it to avoid confusion.

**Comment by lazarevevgeny** [comment]
> Also, I haven't found dependency to KerasRNNInputSlicing and KerasRNNOutputConcatenation to run them after the TF OD API transformations.

Refer to the file `model-optimizer/extensions/front/tf/ObjectDetectionAPI.py`

**Comment by lazarevevgeny** [comment]
> What TensorFlow 2 layer has multiple input slicing and output concatenation pattern in While loop? Maybe you had a chance to look at training code.

It is not a particular layer, it is a hand-written sub-graph performing resize of the input image and also calculating a tensor with shape.

**Review by mvafin — APPROVED** [review]

**Review by pavel-esir — APPROVED** [review]

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py [hunk 6]
```diff
@@ -512,19 +515,104 @@ def update_parameter_shape(graph: Graph, match: [SubgraphMatch, None]):
...
+    This is a anchor transformation which is used to distinguish TF OD API models related transformations.
     """
-    replacement_id = 'ObjectDetectionAPIPreprocessorReplacement'
+    enabled = True
+
+    def run_after(self):
+        return [CropAndResizeReplacement, FakeQuantWithMinMaxVarsToQuantize]
+
+    def find_and_replace_pattern(self, graph: Graph):
+        pass
```
Comment (rkazants) [inline model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:528]: what is a value of this transformation? It is not clear to me how it can distinguish TF OD API models

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py [hunk 7]
```diff
@@ -512,19 +515,104 @@ def update_parameter_shape(graph: Graph, match: [SubgraphMatch, None]):
...
         # PadTFToPad inserts Transpose ops for Pad ops inside the sub-graph corresponding to DetectionOutput.
         # But the inputs corresponding to padding values is re-used as inputs for newly created Pad node. This input
         # is removed during removing nodes from the DO sub-graph so the first input to Transpose is missing which
         # results in TransposeOrderNormalizer transformation failure.
-        return [Pack, TransposeOrderNormalizer, PadTFToPad]
+        return [Pack, TransposeOrderNormalizer, PadTFToPad, SqueezeAxis, StandaloneConstEraser, TFSliceToSliceReplacer,
+                KerasRNNOutputConcatenation, KerasRNNInputSlicing]
+
+    def find_and_replace_pattern(self, graph: Graph):
+        pass
```
Comment (rkazants) [inline model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:549]: I have the same comment since the transformation seems to do nothing.

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py (hunk 6 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:528]: All TF OD API transformations have relationship to the anchor transformations `ObjectDetectionAPITransformationsStart` and `ObjectDetectionAPITransformationsFinish`. Basically all of them are executed after `ObjectDetectionAPITransformationsStart` and before `ObjectDetectionAPITransformationsFinish`. So I grouped all TF OD API transformation into a cluster (refer to the image in the description of the PR). 

TF OD API transformation may leave a graph in a non-normalized state and if some other front transformation will be called in between then this lead to a conversion failure. With the introduction of these anchor transformations we solve this issue and finally we can safely call the graph cleanup to remove dangling nodes after applying all TF OD API transformations after the `ObjectDetectionAPITransformationsFinish`.

We already use similar approach to divide the model conversion pipeline into stages. Refer to the `extensions/front/pass_separator.py` for the example of `FrontStart` and `FrontFinish` anchor transformations.

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py (hunk 7 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:549]: Refer to the comment above.

Comment (lazarevevgeny) [reply model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:549]: And it calls the graph cleanup which is very important :)

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py (hunk 6 above)
Comment (rkazants) [reply model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:528]: I suggest to add this info into the comment

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py [hunk 8]
```diff
@@ -512,19 +515,104 @@ def update_parameter_shape(graph: Graph, match: [SubgraphMatch, None]):
...
+    :return: list of triplets (Node, const_port_index, const_value)
+    """
+    node = first_node.out_port(0).get_destination().node if forward else first_node.in_port(0).get_source().node
+    result = []  # (Node, port # with constant input, value)
+    while node.soft_get('op') in allowed_ops:
+        for port in (0, 1):
+            if node.in_port(port).get_source().node.has_valid('value'):  # this is a constant input to the node
+                result.append((node, port, node.in_port(port).get_source().node.value.copy()))
+                node = node.out_port(0).get_destination().node if forward else node.in_port(1 - port).get_source().node
+                break
```
Comment (rkazants) [inline model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:568]: I think we need to assert to check that node from `allowed_ops` has two input ports for sure, otherwise, passed `allowed_ops` is incorrect. Without this check it can lead to crash in case of incorrect use of this function.

**Review by rkazants — APPROVED** [review]

File: model-optimizer/extensions/front/tf/ObjectDetectionAPI.py (hunk 8 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/front/tf/ObjectDetectionAPI.py:571]: Assert has been added.
